hudi-flink-datasource/hudi-flink/src/main/java/org/apache/hudi/table/HoodieTableSource.java

```diff
@@ -162,7 +163,7 @@ public HoodieTableSource(
     this.limit = limit == null ? NO_LIMIT_CONSTANT : limit;
     this.filters = filters == null ? Collections.emptyList() : filters;
     this.hadoopConf = HadoopConfigurations.getHadoopConf(conf);
-    this.metaClient = StreamerUtil.metaClientForReader(conf, hadoopConf);
+    this.metaClient = Optional.ofNullable(metaClient).orElse(StreamerUtil.metaClientForReader(conf, hadoopConf));
```

Comment:
Another choice is to make `metaClient` and `fileIndex` initialization lazy.

Reply:
> initialization lazy

Here we have some more consideration: 

During planned and optimized **metaclient** or **fileindex** might be used  by different rules' **hoodie source** copy instance for multi times. 
For example, actually **fileindex** is used in **PushPartitionIntoTableSourceScanRule** and **PushFilterIntoSourceScanRuleBase**

So it seems that initialization lazy will also cause the multi construction ?

Reply:
@danny0405 shall we need to do more improvement?

Reply:
@danny0405 are there any other problems for this pr?
Have seen one pr: HUDI-5147 approved to solve he `fileindex` problems, which is part of the contents mentioned above.

Reply:
Can we just decide whether it is null directly instead of using `Optional.ofNullable(metaClient).orElse` here ?

Reply:
> Can we just decide whether it is null directly instead of using `Optional.ofNullable(metaClient).orElse` here ?

@danny0405 updated, pls review it one more time.